File: examples/oshmem_max_reduction.c
/*
 * Copyright (c) 2014      Mellanox Technologies, Inc.
 *                         All rights reserved.
 * $COPYRIGHT$
 *
 * Additional copyrights may follow
 *
 * $HEADER$
 *
 * reduce [0,1,2] + _my_pe() across 4 PEs with MAX()
 */

#include <stdio.h>
#include <string.h>

#include <shmem.h>

long pSync[_SHMEM_BCAST_SYNC_SIZE];

#define N 3

long src[N];
long dst[N];
long pWrk[_SHMEM_REDUCE_SYNC_SIZE];

int  main(void)
{
    int i;
    int my_pe, num_pes;

    for (i = 0; i < SHMEM_BCAST_SYNC_SIZE; i += 1) {
        pSync[i] = _SHMEM_SYNC_VALUE;
    }

    start_pes(0);

    my_pe = _my_pe();
    num_pes = _num_pes();

    for (i = 0; i < N; i += 1) {
        src[i] = my_pe + i;
    }

    shmem_barrier_all();

    shmem_long_max_to_all(dst, src, N, 0, 0, num_pes, pWrk, pSync);

    printf("%d/%d dst =", my_pe, num_pes);

    for (i = 0; i < N; i+= 1) {
        printf(" %ld", dst[i]);
    }

    printf("\n");

    return 0;
}

File: examples/oshmem_max_reduction.java
import shmem.*;
import java.nio.*;

public class oshmem_max_reduction
{
    private static final int N = 3;

    public static void main(String[] args) throws ShMemException
    {
        ShMem.startPEs(0);

        int numPEs = ShMem.getNumPEs(),
            myPE   = ShMem.getMyPE();

        Addr  src   = new Addr(8 * N), // long is 8 bytes.
              dst   = new Addr(8 * N),
              pWrk  = new Addr(8 * ShMem.REDUCE_SYNC_SIZE);
        PSync pSync = new PSync(ShMem.BCAST_SYNC_SIZE);

        LongBuffer srcBuf = src.asLongBuffer(),
                   dstBuf = dst.asLongBuffer();

        for(int i = 0; i < N; i++)
            srcBuf.put(i, myPE + i);

        ShMem.barrierAll();
        dst.maxToAllLong(src, N, 0, 0, numPEs, pWrk, pSync);

        StringBuilder sb = new StringBuilder();
        sb.append(myPE +"/"+ numPEs +" dst =");

        for(int i = 0; i < N; i++)
            sb.append(" "+ dstBuf.get(i));

        sb.append("\n");
        System.out.print(sb);

        src.free();
        dst.free();
        pWrk.free();
        pSync.free();
    }
}
